### PR TITLE
Added micronucleus

### PR DESCRIPTION
**Pull request by cab404**
###### Motivation for this change

Micronucleus is a nice bootloader for ATTiny boards, and is used by Digispark. This is a command line tool to flash stuff into it.

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [X] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [X] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by cab404** [comment]
Up?

File: pkgs/development/tools/misc/micronucleus/default.nix [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+   patchPhase = ''
+     sed -i 's\/usr/local/bin\'$out/bin/'\g' Makefile
+   '';
+
+   installPhase = ''
+     mkdir -p $out/bin
+     mkdir -p $out/lib/udev
+     cp micronucleus $out/bin
+     cp 49-micronucleus.rules $out/lib/udev
```
Comment (flokli) [inline pkgs/development/tools/misc/micronucleus/default.nix]: Future work, not needed for this PR:

You might want to PR a makefile change upstream, which updates their install taget to use `DESTDIR` and installs the udev rule too.

We'd then only need to set `makeFlags = [ "DESTDIR=$(out)" ];`, and could drop the custom `installPhase` alltogether for the next release.

**Review by flokli — COMMENTED** [review]
Thanks for the contribution, and stepping up as a maintainer! Some comments.

**Review by flokli — CHANGES_REQUESTED** [review]
Also, format your commit messages according to `CONTRIBUTING.md`.

**Comment by cab404** [comment]
> Also, format your commit messages according to `CONTRIBUTING.md`.
Done

**Comment by flokli** [comment]
No, the second commit should be `micronucleus: init at 2.04`.

**Comment by cab404** [comment]
Done.

On January 1, 2020 11:53:53 PM GMT+01:00, Florian Klink <notifications@github.com> wrote:
>No, the second commit should be `micronucleus: init at 2.04`.
>
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/NixOS/nixpkgs/pull/76621#issuecomment-570090317
—
cab

**Comment by flokli** [comment]
@GrahamcOfBorg build micronucleus

**Comment by flokli** [comment]
Build fails on darwin, because upstreams makefile hardcodes `CC` to `gcc`, which doesn't exist there.

You could try setting `CC = if stdenv.isDarwin then "clang" else "cc";` if that's all to get it working, or add remove darwin from `meta.platforms`.

**Comment by cab404** [comment]
> Build fails on darwin, because upstreams makefile hardcodes `CC` to `gcc`, which doesn't exist there.
> 
> You could try setting `CC = if stdenv.isDarwin then "clang" else "cc";` if that's all to get it working, or add remove darwin from `meta.platforms`.

can i just use clang then?

**Comment by cab404** [comment]
@GrahamcOfBorg build micronucleus

**Comment by flokli** [comment]
@GrahamcOfBorg build micronucleus

**Review by flokli — APPROVED** [review]

**Pull request merged**